File: C++/207_Course_Schedule.cpp
class Solution {
public:
    unordered_set<int> parent;
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> graph;
        for(int i = 0; i < prerequisites.size(); i++){
            graph[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        for(int i = 0; i < numCourses; i++){
            if(parent.count(i) == 0){
                parent.insert(i);
                if(dfs(i, graph)) return false;
                parent.erase(i);
            }
        }
        return true;
    }
    bool dfs(int s, unordered_map<int, vector<int>>& graph){
        for(int x : graph[s]){
            if(parent.count(x) == 1){
                return true;
            } else {
                parent.insert(x);
                if(dfs(x, graph)) return true;
                parent.erase(x);
            }
        }
        return false;
    }
};

File: Python/207_Course_Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = {}
        parent = {}
        for i in range(numCourses):
            graph[i] = []
        for edge in prerequisites:
            graph[edge[1]].append(edge[0])
        return self.DFS(graph, parent)
    
    def DFS(self, graph, parent):
        for s in graph.keys():
            if s not in parent:
                parent[s] = s
                a = self.DFS_Visit(s, graph, parent)
                del parent[s]
                if a:
                    return False
        return True
    
    def DFS_Visit(self, s, graph, parent):
        for v in graph[s]:
            if v in parent:
                return True
            else:
                parent[v] = s
                a = self.DFS_Visit(v, graph, parent)
                if a:
                    return True
                del parent[v]

    